### PR TITLE
Fix MSan error in "rdkafka" library #12990

### LINKED ISSUE #12990: MemorySanitizer: use-of-uninitialized-value test_storage_kafka
**How to reproduce**
* 20.7(master)
* Memory sanitizer
* Integration test test_storage_kafka
```
Uninitialized bytes in __interceptor_getaddrinfo at offset 20 inside [0x7f7b39a07300, 48)
==1==WARNING: MemorySanitizer: use-of-uninitialized-value
==1==WARNING: invalid path to external symbolizer!
==1==WARNING: Failed to use and restart external symbolizer!
    #0 0x3259c51f  (/usr/bin/clickhouse+0x3259c51f)
    #1 0x32560452  (/usr/bin/clickhouse+0x32560452)
    #2 0x3288b54c  (/usr/bin/clickhouse+0x3288b54c)
    #3 0x7f7c16816668  (/lib/x86_64-linux-gnu/libpthread.so.0+0x9668)
    #4 0x7f7c1672d322  (/lib/x86_64-linux-gnu/libc.so.6+0x122322)

  Uninitialized value was created by an allocation of 'hints' in the stack frame of function 'rd_getaddrinfo'
    #0 0x3259c150  (/usr/bin/clickhouse+0x3259c150)

SUMMARY: MemorySanitizer: use-of-uninitialized-value (/usr/bin/clickhouse+0x3259c51f)

addr2line -afiCe ~/memory-ch/output/binary/clickhouse-server 0x3259c51f 0x32560452 0x3288b54c 0x9668 0x122322 0x3259c150 0x3259c51f
0x000000003259c51f
rd_getaddrinfo
/build/obj-x86_64-linux-gnu/../contrib/librdkafka/src/rdaddr.c:168
0x0000000032560452
rd_kafka_broker_resolve
/build/obj-x86_64-linux-gnu/../contrib/librdkafka/src/rdkafka_broker.c:840
rd_kafka_broker_connect
/build/obj-x86_64-linux-gnu/../contrib/librdkafka/src/rdkafka_broker.c:1890
rd_kafka_broker_thread_main
/build/obj-x86_64-linux-gnu/../contrib/librdkafka/src/rdkafka_broker.c:4914
0x000000003288b54c
_thrd_wrapper_function
/build/obj-x86_64-linux-gnu/../contrib/librdkafka/src/tinycthread.c:576
0x0000000000009668
??
??:0
0x0000000000122322
??
??:0
0x000000003259c150
rd_getaddrinfo
/build/obj-x86_64-linux-gnu/../contrib/librdkafka/src/rdaddr.c:149
0x000000003259c51f
rd_getaddrinfo
/build/obj-x86_64-linux-gnu/../contrib/librdkafka/src/rdaddr.c:168
```

### PR DESCRIPTION
**Pull request by alexey-milovidov**
Changelog category (leave one):
- Build/Testing/Packaging Improvement

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Fix MSan error in "rdkafka" library. This closes #12990. Updated `rdkafka` to version 1.5 (master).

### CODE REVIEW

**Review by filimonov — APPROVED** [review]
It also updates librdkafka to version 1.5.0 (that's completely ok, but should be mentioned in changelog). 

Pushing the fix to upstream in desirable, rdkafka is quite an active project and there is quite a lot of development happening there recently, we don't want to stuck with 1.5 or reapply the patch every time.

**Comment by alexey-milovidov** [comment]
@filimonov
Something was changed in Kafka so the integration test failed.

**Comment by alexey-milovidov** [comment]
Most of Kafka testa are Ok, only single case has failed. Looks like something trivial...

**Comment by alexey-milovidov** [comment]
@azat There are strange log messages:
```
2020.07.28 18:03:53.256298 [ 54 ] {6e2c65f5-f212-4b29-88df-076e5ee755a6} <Trace> StorageKafka (kafka): Already subscribed to topics: []
2020.07.28 18:03:53.256343 [ 54 ] {6e2c65f5-f212-4b29-88df-076e5ee755a6} <Trace> StorageKafka (kafka): Already assigned to: [  ]
```

Double whitespace inside brackets.

**Comment by alexey-milovidov** [comment]
`Offset position: [  ]`

Also double whitespace.

**Comment by alexey-milovidov** [comment]
```
2020.07.28 18:03:53.757221 [ 54 ] {6e2c65f5-f212-4b29-88df-076e5ee755a6} <Trace> StorageKafka (kafka): Polled batch of 1 messages. Offset position: [  ]
2020.07.28 18:03:53.757275 [ 54 ] {6e2c65f5-f212-4b29-88df-076e5ee755a6} <Error> StorageKafka (kafka): Consumer error: Broker: Unknown topic or partition
2020.07.28 18:03:53.757306 [ 54 ] {6e2c65f5-f212-4b29-88df-076e5ee755a6} <Error> StorageKafka (kafka): There were 18446744073709551615 messages with an error
2020.07.28 18:03:53.757316 [ 54 ] {6e2c65f5-f212-4b29-88df-076e5ee755a6} <Error> StorageKafka (kafka): Only errors left
```

How to interpret this?

**Comment by alexey-milovidov** [comment]
@azat you have trivial bug in code.

**Comment by alexey-milovidov** [comment]
What does it mean?

```
2020.07.28 18:03:53.266388 [ 51 ] {} <Debug> StorageKafka (kafka): [rdk:CGRPSTATE] [thrd:main]: Group "flush_by_block_size" changed state wait-broker-transport -> up (v1, join-state init)
2020.07.28 18:03:53.266402 [ 51 ] {} <Debug> StorageKafka (kafka): [rdk:JOIN] [thrd:main]: Group "flush_by_block_size": join with 0 (1) subscribed topic(s)
2020.07.28 18:03:53.266415 [ 51 ] {} <Debug> StorageKafka (kafka): [rdk:CGRPMETADATA] [thrd:main]: consumer join: metadata for subscription only available for 0/1 topics (-1ms old)
2020.07.28 18:03:53.266431 [ 51 ] {} <Debug> StorageKafka (kafka): [rdk:JOIN] [thrd:main]: Group "flush_by_block_size": postponing join until up-to-date metadata is available
2020.07.28 18:03:53.266471 [ 53 ] {} <Debug> StorageKafka (kafka): [rdk:SEND] [thrd:kafka1:19092/bootstrap]: kafka1:19092/1: Sent MetadataRequest (v4, 70 bytes @ 0, CorrId 6)
2020.07.28 18:03:53.268119 [ 53 ] {} <Debug> StorageKafka (kafka): [rdk:RECV] [thrd:kafka1:19092/bootstrap]: kafka1:19092/1: Received MetadataResponse (v4, 86 bytes, CorrId 6, rtt 1.63ms)
2020.07.28 18:03:53.268147 [ 51 ] {} <Debug> StorageKafka (kafka): [rdk:METADATA] [thrd:main]: kafka1:19092/1: 1/1 requested topic(s) seen in metadata
2020.07.28 18:03:53.268163 [ 51 ] {} <Debug> StorageKafka (kafka): [rdk:TOPICERR] [thrd:main]: Subscribed topic not available: flush_by_block_size: Broker: Unknown topic or partition
```

**Comment by azat** [comment]
>Double whitespace inside brackets.

@alexey-milovidov this goes from cppkafka:

https://github.com/ClickHouse-Extras/cppkafka/blob/f555ee36aaa74d17ca0dab3ce472070a610b2966/src/topic_partition_list.cpp#L117-L127

**Comment by azat** [comment]
>@azat you have trivial bug in code.

@alexey-milovidov Indeed, thank you for the fix!

File: tests/integration/test_storage_kafka/test.py [hunk 1]
```diff
@@ -1535,7 +1537,9 @@ def test_kafka_flush_by_block_size(kafka_cluster):
         messages.append(json.dumps({'key': 0, 'value': 0}))
     kafka_produce('flush_by_block_size', messages)
 
-    time.sleep(1)
+    # Wait for Kafka engine to consume this data
+    while 1 != int(instance.query("SELECT count() FROM system.parts WHERE database = 'test' AND table = 'view' AND name = 'all_1_1_0'")):
```
Comment (azat) [inline tests/integration/test_storage_kafka/test.py:1541]: worth limit number of attempts?

**Comment by azat** [comment]
>What does it mean?
> `2020.07.28 18:03:53.268163 [ 51 ] {} <Debug> StorageKafka (kafka): [rdk:TOPICERR] [thrd:main]: Subscribed topic not available: flush_by_block_size: Broker: Unknown topic or partition`

I guess that the topic hasn't been created yet.

**Comment by alexey-milovidov** [comment]
I read the logs of previous version and looks like that it first starting to argue that the topic is not created and then subscribe and continue to run. On new version it does not read data from Kafka.

File: tests/integration/test_storage_kafka/test.py (hunk 1 above)
Comment (alexey-milovidov) [reply tests/integration/test_storage_kafka/test.py:1541]: Maybe but they are already limited by `@pytest.mark.timeout(600)`

**Review by azat — COMMENTED** [review]
> alexey-milovidov added the help wanted label

@alexey-milovidov anything left? LGTM

**Comment by alexey-milovidov** [comment]
@azat It's strange - I just added retries and test start to pass.
I did not suppose that it will fix the test.

Qustions:
- if test is unstable (depends on race conditions), why it was passing in previous version of rdkafka?
- if we had to add retries on new version of rdkafka, does it mean that we have some performance degradation?

**Comment by alexey-milovidov** [comment]
Nevertheless, we also have to fix Stress test (memory) and AST-based query fuzzer.

**Comment by azat** [comment]
>Stress test (memory) — Killed by signal (in clickhouse-server.log)

Introduced in #12666

**Comment by azat** [comment]
>AST-based query fuzzer — Fuzzer failed (134). See the logs

No fatal errors in the log (and lack of the server log)

**Comment by alexey-milovidov** [comment]
Ok. As all the errors are investigated let's merge.

> No fatal errors in the log (and lack of the server log)

We will ask @akuzm to improve usability of the output.
I also did not find any erroneous queries in the log.

**Comment by azat** [comment]
>Ok. As all the errors are investigated let's merge.

I was going to look into the test a little bit deeper (why it fails after upgrading librdkafka), will do this anyway

**Comment by filimonov** [comment]
Actually retries breaks the intentions of that test. It should pass w/o retries.

**Comment by azat** [comment]
>Actually retries breaks the intentions of that test. It should pass w/o retries.

@filimonov are you looking into this?

**Comment by filimonov** [comment]
Looking on that. 

Most probably it's caused by that change: 'Subscribing to non-existent and unauthorized topics will now propagate errors ... Consumer will no longer trigger auto creation of topics'. Will try to create topic explicitly...

**Comment by filimonov** [comment]
https://github.com/ClickHouse/ClickHouse/pull/13285

**Comment by qoega** [comment]
Still getting an error on master
https://clickhouse-test-reports.s3.yandex.net/0/f7edcdf7c847f7d45ceede4a6073400ec2c985b7/integration_tests_(memory).html
```
Uninitialized bytes in __interceptor_getaddrinfo at offset 20 inside [0x7febe03d1ec0, 48)
==7==WARNING: MemorySanitizer: use-of-uninitialized-value
    #0 0x3bcc351d in get_fqhostname obj-x86_64-linux-gnu/../contrib/cyrus-sasl/lib/saslutil.c:548:9
    #1 0x3bcf6e7f in sasl_client_new obj-x86_64-linux-gnu/../contrib/cyrus-sasl/lib/client.c:515:7
    #2 0x3bb91c19 in rd_kafka_sasl_cyrus_client_new obj-x86_64-linux-gnu/../contrib/librdkafka/src/rdkafka_sasl_cyrus.c:500:13
    #3 0x3bacfd9d in rd_kafka_sasl_client_new obj-x86_64-linux-gnu/../contrib/librdkafka/src/rdkafka_sasl.c:265:13
    #4 0x3b689f83 in rd_kafka_broker_connect_auth obj-x86_64-linux-gnu/../contrib/librdkafka/src/rdkafka_broker.c:2263:8
    #5 0x3b6a4075 in rd_kafka_broker_handle_SaslHandshake obj-x86_64-linux-gnu/../contrib/librdkafka/src/rdkafka_broker.c:2199:2
    #6 0x3b70c8e3 in rd_kafka_buf_callback obj-x86_64-linux-gnu/../contrib/librdkafka/src/rdkafka_buf.c:480:17
    #7 0x3b681189 in rd_kafka_req_response obj-x86_64-linux-gnu/../contrib/librdkafka/src/rdkafka_broker.c:1775:9
    #8 0x3b681189 in rd_kafka_recv obj-x86_64-linux-gnu/../contrib/librdkafka/src/rdkafka_broker.c:1893:3
    #9 0x3bb18ff2 in rd_kafka_transport_io_event obj-x86_64-linux-gnu/../contrib/librdkafka/src/rdkafka_transport.c:742:11
    #10 0x3bb18ff2 in rd_kafka_transport_io_serve obj-x86_64-linux-gnu/../contrib/librdkafka/src/rdkafka_transport.c:801:17
    #11 0x3b6bd7b3 in rd_kafka_broker_ops_io_serve obj-x86_64-linux-gnu/../contrib/librdkafka/src/rdkafka_broker.c:3380:21
    #12 0x3b6adb2d in rd_kafka_broker_consumer_serve obj-x86_64-linux-gnu/../contrib/librdkafka/src/rdkafka_broker.c:4961:17
    #13 0x3b6adb2d in rd_kafka_broker_serve obj-x86_64-linux-gnu/../contrib/librdkafka/src/rdkafka_broker.c:5066:17
    #14 0x3b694799 in rd_kafka_broker_thread_main obj-x86_64-linux-gnu/../contrib/librdkafka/src/rdkafka_broker.c:5208:25
    #15 0x3bb86676 in _thrd_wrapper_function obj-x86_64-linux-gnu/../contrib/librdkafka/src/tinycthread.c:576:9
    #16 0x7fecf4eaf608 in start_thread (/lib/x86_64-linux-gnu/libpthread.so.0+0x9608)
    #17 0x7fecf4dd6292 in clone (/lib/x86_64-linux-gnu/libc.so.6+0x122292)

  Uninitialized value was created by an allocation of 'hints' in the stack frame of function 'get_fqhostname'
    #0 0x3bcc31b0 in get_fqhostname obj-x86_64-linux-gnu/../contrib/cyrus-sasl/lib/saslutil.c:523

SUMMARY: MemorySanitizer: use-of-uninitialized-value obj-x86_64-linux-gnu/../contrib/cyrus-sasl/lib/saslutil.c:548:9 in get_fqhostname
```

**Comment by azat** [comment]
@qoega seems that you already fixed it (https://github.com/ClickHouse-Extras/cyrus-sasl/commit/e6466edfd638cc5073debe941c53345b18a09512) are you going to update the submodule?

**Pull request merged**